src/utils/workerPool.ts

```diff
@@ -0,0 +1,143 @@
+import { CancellationToken, Progress } from "vscode";
+import { logError } from "../errors";
+import { Logger } from "../logging";
+
+const logger = new Logger("workerPool");
+
+/** Result of a worker pool execution containing either a result or error. */
+export interface ExecutionResult<R> {
+  result?: R;
+  error?: Error;
+}
+
+/** {@link ExecutionResult} that always includes a `result` and never an `error`. */
+export interface SuccessResult<R> extends ExecutionResult<R> {
+  result: R;
+  error: undefined;
+}
+
+/** {@link ExecutionResult} that always includes an `error` and never a `result`. */
+export interface ErrorResult extends ExecutionResult<unknown> {
+  result: undefined;
+  error: Error;
+}
+
+/** Type guard to check if an {@link ExecutionResult} is a {@link SuccessResult}. */
+export function isSuccessResult<R>(
+  result: ExecutionResult<R> | undefined,
+): result is SuccessResult<R> {
+  return result?.result !== undefined;
+}
+
+/** Type guard to check if an {@link ExecutionResult} is an {@link ErrorResult}. */
+export function isErrorResult(result: ExecutionResult<unknown> | undefined): result is ErrorResult {
+  return result?.error !== undefined;
+}
+
+interface WorkerPoolOptions {
+  maxWorkers: number;
+  taskName?: string;
+}
+
+/**
+ * Execute a callback against an array of (non-`undefined`) items using a worker pool pattern.
+ * Results will be returned in the same order as the input items, with each result containing either
+ * a result or an error.
+ *
+ * @param items Array of items to process
+ * @param callable Async function to execute for each item
+ * @param options Worker pool options
+ * @param progress Optional progress reporter
+ * @param token Optional cancellation token
+ * @returns Array of results, each containing either a result or error
+ */
+export async function executeInWorkerPool<T, R>(
```

Comment:
My gut feel is that the callable should come first. It seems to be the more important of the params, as in spirit of `apply(fxn, data)`.

Reply:
Done in https://github.com/confluentinc/vscode/pull/992/commits/bea0a0f89d206a37bf063ae6055e7c2e1d52f2b8